dali/pipeline/executor/executor2/exec_graph.cc

```diff
@@ -0,0 +1,293 @@
+      ws->AddInput(std::shared_ptr<TensorList<GPUBackend>>(nullptr));
+    } else {
+      ws->AddInput(std::shared_ptr<TensorList<CPUBackend>>(nullptr));
+    }
+  }
+  for (int i = 0, nout = outputs.size(); i < nout; i++) {
+    bool gpu = outputs[i].device == StorageDevice::GPU;
+    if (gpu) {
+      ws->AddOutput(std::shared_ptr<TensorList<GPUBackend>>(nullptr));
+    } else {
+      ws->AddOutput(std::shared_ptr<TensorList<CPUBackend>>(nullptr));
+    }
+  }
+  return ws;
+}
+
+/** Obtains a worskpace from a workspace cache or, if not found, creates a new one. */
+std::pair<std::unique_ptr<Workspace>, SharedEventLease> ExecNode::GetWorkspace(WorkspaceParams params) {
+  if (!ws_) {
+    assert(!has_workspace_);
+    if (op) {
+      ws_ = CreateOpWorkspace();
+    } else {
+      ws_ = CreateOutputWorkspace();
+    }
+    has_workspace_ = true;
+  }
+  if (!params.env)
+    params.env = &env;
+
+  ws_event_.reset();
+  ws_->set_event(nullptr);
+
+  for (int o = 0; o < ws_->NumOutput(); o++) {
+    if (ws_->OutputIsType<GPUBackend>(o)) {
+      if (!ws_event_)
+        ws_event_ = SharedEventLease::Get();
+      ws_->set_event(ws_event_);
+      break;
+    }
+  }
+
+  ApplyWorkspaceParams(*ws_, params);
+  return { std::move(ws_), ws_event_ };
+}
+
+void ExecNode::PutWorkspace(std::unique_ptr<Workspace> &&ws) {
+  assert(has_workspace_);
+  assert(ws);
+  assert(!ws_);
+  assert(!ws->has_event() || ws->event() == ws_event_.get());
+  ws_ = std::move(ws);
+}
+
+void ExecNode::AddDataDeps() {
+  for (auto &edge : inputs) {
+    assert(edge->producer->main_task_);
+    main_task_->Subscribe(edge->producer->main_task_, edge->producer_output_idx);
+  }
+}
+
+void ExecGraph::PrepareIteration(const WorkspaceParams &params) {
+  if (nodes_.empty())
+    throw std::logic_error("Cannot execute an empty graph");
+
+  if (params.batch_size <= 0)
+    throw std::runtime_error("Batch size must be a positive number.");
+
+  Sort();
+  Validate();
+  Analyze();
+
+  // Create a special task that checks the predicted batch size
+  // and populates the respective field in IterationData
+  auto prev_infer = infer_batch_size_task_;
+  infer_batch_size_task_ = InferBatchSize(params.iter_data, params.batch_size);
+  if (infer_batch_size_task_) {
+    if (prev_infer)
+      infer_batch_size_task_->Succeed(prev_infer);
+  } else {
+    params.iter_data->default_batch_size = params.batch_size;
+  }
+
+  for (auto &n : nodes_) {
+    n.NextIter();
+    n.CreateMainTask(params);
+  }
+  for (auto &n : nodes_) {
+    n.AddDataDeps();
+    n.CreateAuxTasks();
+    // all root tasks must succeed batch size inference
+    if (n.inputs.empty() && infer_batch_size_task_)
+      n.main_task_->Subscribe(infer_batch_size_task_);
+  }
+}
+
+inline std::string_view NodeName(const ExecNode &n) {
+  return !n.instance_name.empty() ? n.instance_name : n.op->GetSpec().SchemaName();
+}
+
+tasking::SharedTask ExecGraph::InferBatchSize(const std::shared_ptr<IterationData> &iter_data,
```

Comment:
As already commented, this could be a third type of OpTask, giving even more reasons to split the OpTask implementation and have a group of 3 (or more in the future) OpTasks that are submittable to the exec_graph.

Reply:
Well, it could be _a task_, but not OpTask - just like OutputTask wouldn't be an OpTask.
OpTask (and the hypothetical OutputTask) could have a common base (ExecNodeTask?) - but this particular task isn't one. Batch size inference isn't an ExecNode and I don't think it should be, because there's no good way to represent the dependency between it and other tasks (normally those correspond strictly to taking o-th output as i-th input).

Reply:
Done - ExecNodeTask is a base for OpTask and OutputTask. The latter two are hidden inside .cc file.